### ISSUE TITLE
Dependency missing in installation from source and pytest segmentation fault

### ISSUE THREAD

**Issue by ojeda-e** [deepchem/deepchem#3002]
## ❓ Questions & Help
Hi, since I am interested in contributing to DeepChem, I followed the steps provided in the documentation to install DeepChem [from source with Conda](https://deepchem.readthedocs.io/en/latest/get_started/installation.html#from-source-with-conda). Then, to verify my installation was successful, I proceeded to run the tests. This is the output
```
~/python_playground/deepchem (master) » pytest -m "not slow" deepchem                                                                                                              estefania@BioMac
======================================================================================= test session starts ========================================================================================
platform darwin -- Python 3.8.13, pytest-7.1.2, pluggy-1.0.0
rootdir: /Users/estefania/python_playground/deepchem, configfile: setup.cfg
plugins: flaky-3.7.0, typeguard-2.13.3, cov-3.0.0
collected 870 items / 1 error / 34 deselected / 836 selected                                                                                                                                       

============================================================================================== ERRORS ==============================================================================================
___________________________________________________________ ERROR collecting deepchem/models/lightning/tests/test_dc_lightning_module.py ___________________________________________________________
deepchem/models/lightning/tests/test_dc_lightning_module.py:9: in <module>
    from deepchem.models.lightning.dc_lightning_module import DCLightningModule
deepchem/models/lightning/__init__.py:2: in <module>
    from deepchem.models.lightning.dc_lightning_module import DCLightningModule
deepchem/models/lightning/dc_lightning_module.py:9: in <module>
    class DCLightningModule(pl.LightningModule):
E   NameError: name 'pl' is not defined
```

Checking at `deepchem/models/lightning/dc_lightning_module.py`, seems like the dependency `pytorch-lightening` is missing.
I proceeded to install the missing dependency with `pip install pytorch-lightning`. 

After installation, when running `pytest -m "not slow" deepchem` collection of all tests works as expected. However, there is a segmentation fault in `deepchem/feat/tests/test_smiles_tokenizer.py`, as shown in the image below:

<img width="1378" alt="image" src="https://user-images.githubusercontent.com/27988163/180037056-f1ad25a5-0d24-4b74-adf9-4fb7b16e1bf6.png">

Questions:
- Shouldn't be `pytorch-lightning` included in `requirements/torch/env_torch.mac.cpu.yml`?
- Would you please point to the workflow that runs the CI tests on macOS? I only found [this](https://github.com/deepchem/deepchem/blob/master/.github/workflows/torch_setup.yml#L36) but is running on Linux and Windows only.
- Is there an alternative way to run the tests to pass locally?   

Thanks.

**Comment by arunppsg** [comment]
>     * Shouldn't be `pytorch-lightning` included in `requirements/torch/env_torch.mac.cpu.yml`?

Yes, pytorch lightning should be added to `env_torch.mac.cpu.yml`. Will you be interested in making a PR for the same?
 
>     * Would you please point to the workflow that runs the CI tests on macOS? I only found [this](https://github.com/deepchem/deepchem/blob/master/.github/workflows/torch_setup.yml#L36) but is running on Linux and Windows only.
Currently, we are testing only on linux and windows. We have to add ci tests on macos. Nevertheless, you can develop deepchem in macos. You can look at [good first issues](https://github.com/deepchem/deepchem/issues?q=is%3Aopen+is%3Aissue+label%3A%22Good+First+Contribution%22) for ideas.


>     * Is there an alternative way to run the tests to pass locally?
In case of `test_smiles_tokenizer.py`, I noticed that the test is passing in our CI and also locally. Have you installed `torch` in your environment?

**Comment by ojeda-e** [comment]
Hi @arunppsg, thanks for your answer.
Yes, I'm interested in opening the PR to add the missing dependency.

Thanks for clarifying, I'll give it a try on Linux today. 

As for, `test_smiles_tokenizer.py`, I do have installed torch. I have:
```
(deepchem)  /deepchem (master) » pip list | grep torch
pytorch-lightning            1.6.5
torch                        1.10.1
torch-geometric              2.0.4
torch-scatter                2.0.9
torch-sparse                 0.6.14
torchmetrics                 0.9.2
```
I'll try to reproduce the same result (segmentation fault when running `test_smiles_tokenizer.py`) on Linux. I'll keep this thread updated.

**Comment by ojeda-e** [comment]
Sorry for the late reply, @arunppsg.

**TL;DR:** The dependency `libstdcxx-ng=12.1.0` is needed to run tests in deepchem env on Ubuntu. Three unit tests fail.

I tried installing from source on Linux [as shown in docs](https://deepchem.readthedocs.io/en/latest/get_started/installation.html#from-source-with-conda) in two different distributions:
- pop_OS! 20.04 LTS
- Ubuntu 22.04

Please note that the output reported below was identical on both distributions (which are also separate hardware). After installing the instructions provided in docs, I ran tests with `pytest -m "not slow" deepchem`. I found 3 warnings and 116 errors. The tests weren't even collected. The errors (all of them) show as:

```
E   ImportError: /home/estefania/anaconda3/envs/deepchem/lib/python3.8/site-packages/numexpr/../../../libstdc++.so.6: version `GLIBCXX_3.4.30' not found (required by /home/estefania/anaconda3/envs/deepchem/lib/python3.8/site-packages/scipy/fft/_pocketfft/pypocketfft.cpython-38-x86_64-linux-gnu.so)
```
 
I tried to find the installed versions of `GLIBCXX_3.4.` with:

```
 strings  /home/estefania/anaconda3/envs/deepchem/lib/python3.8/site-packages/numexpr/../../../libstdc++.so.6 | grep "^GLIBCXX_3.4."
``` 
 
The output shows all the versions, but 3.4.** doesn't exist:

```
GLIBCXX_3.4
GLIBCXX_3.4.1
GLIBCXX_3.4.2
GLIBCXX_3.4.3
GLIBCXX_3.4.4
GLIBCXX_3.4.5
GLIBCXX_3.4.6
GLIBCXX_3.4.7
GLIBCXX_3.4.8
GLIBCXX_3.4.9
GLIBCXX_3.4.10
GLIBCXX_3.4.11
GLIBCXX_3.4.12
GLIBCXX_3.4.13
GLIBCXX_3.4.14
GLIBCXX_3.4.15
GLIBCXX_3.4.16
GLIBCXX_3.4.17
GLIBCXX_3.4.18
GLIBCXX_3.4.19
GLIBCXX_3.4.20
GLIBCXX_3.4.21
GLIBCXX_3.4.22
GLIBCXX_3.4.23
GLIBCXX_3.4.24
GLIBCXX_3.4.25
GLIBCXX_3.4.26
GLIBCXX_3.4.27
GLIBCXX_3.4.28
GLIBCXX_3.4.29
 ```

I found a possible solution by installing the gcc lib with conda (see [this link](https://stackoverflow.com/questions/48453497/anaconda-libstdc-so-6-version-glibcxx-3-4-20-not-found)):
```
conda install -c anaconda gcclib
```
but came to see that doesn't fix the issue: when running tests, the error is the same:
```
E   ImportError: /home/estefania/anaconda3/envs/deepchem/lib/python3.8/site-packages/numexpr/../../../libstdc++.so.6: version `GLIBCXX_3.4.30' not found (required by /home/estefania/anaconda3/envs/deepchem/lib/python3.8/site-packages/scipy/fft/_pocketfft/pypocketfft.cpython-38-x86_64-linux-gnu.so)
```

Checking the location of the file mentioned in the error it shows:

```
~/python_playground/deepchem (master) » ls  /home/estefania/anaconda3/envs/deepchem/lib/python3.8/site-packages/numexpr/../../../libstdc++*                                        

/home/estefania/anaconda3/envs/deepchem/lib/python3.8/site-packages/numexpr/../../../libstdc++.so -> libstdc++.so.6.0.29

/home/estefania/anaconda3/envs/deepchem/lib/python3.8/site-packages/numexpr/../../../libstdc++.so.6 -> libstdc++.so.6.0.29

/home/estefania/anaconda3/envs/deepchem/lib/python3.8/site-packages/numexpr/../../../libstdc++.so.6.0.21 -> libstdc++.so.6.0.24

/home/estefania/anaconda3/envs/deepchem/lib/python3.8/site-packages/numexpr/../../../libstdc++.so.6.0.29
```
which essentially means that `libstdc++.so.` exists, just not the version expected by scipy.

I checked the versions of gcc installed by the `deepchem` environment provided in the installation instruction and it shows:
```
~/python_playground/deepchem (master) » conda list | grep "gcc"             
_libgcc_mutex             0.1                 conda_forge    conda-forge
libgcc                    7.2.0                h69d50b8_2    anaconda
libgcc-ng                 12.1.0              h8d9b700_16    conda-forge
~/python_playground/deepchem (master) » conda list | grep "libstd"
libstdcxx-ng              11.2.0               h1234567_1
```
However,  according to the [libstdcxx-ng conda page](https://anaconda.org/conda-forge/libstdcxx-ng), the latest version of `libstdcxx-ng` is 12.1.0, not 11.2.0 as installed in the `deepchem` environment with `source scripts/install_deepchem_conda.sh 3.8 cpu`.

I created a new environment forcing `libstdcxx-ng=12.1.0` and now it collects all the tests. However, there are three tests still failing:
```
FAILED deepchem/models/tests/test_cnn.py::TestCNN::test_residual_cnn_classification - assert 0.875 > 0.9

FAILED deepchem/models/tests/test_megnet.py::test_megnet_classification - RuntimeError: Sizes of tensors must match except in dimension 1. Expected size 160 but got size 159 for tensor number 1 in the list.

FAILED deepchem/models/tests/test_mpnn.py::test_mpnn_regression - assert 0.5587267067766437 < 0.5
```

Although not perfect, this is some progress already. But in any case, according to what I am experiencing in the two Linux distributions, I do need to install `libstdcxx-ng=12.1.0` to make it work. I haven't really checked the CI in full detail. It's a bit puzzling to know that tests pass in the CI but not locally when installing from source. My guess is that one of the dependencies installed in the CI has a dependency that uses the correct version of `libstdcxx-ng`, then making the tests pass, but not sure about that.
Thanks!

**Comment by arunppsg** [comment]
Sounds good! Of the test failures, the first and third are flaky failures. I am not able to reproduce the second failure. Can you run `pytest deepchem/models/tests/test_megnet.py` and check for the error?

**Comment by ojeda-e** [comment]
I found that the test fails when I am running the whole test suite only.

First, I ran `deepchem/models/tests/test_megnet.py` only and I obtained:
```
~/python_playground/deepchem (master*) » pytest deepchem/models/tests/test_megnet.py -vv                                                                                        estefania@ada
==================================================================================== test session starts =====================================================================================
platform linux -- Python 3.8.13, pytest-7.1.2, pluggy-1.0.0 -- /home/estefania/anaconda3/envs/deepchem/bin/python
cachedir: .pytest_cache
rootdir: /home/estefania/python_playground/deepchem, configfile: setup.cfg
plugins: flaky-3.7.0, cov-3.0.0, typeguard-2.13.3
collected 4 items                                                                                                                                                                            

deepchem/models/tests/test_megnet.py::test_megnet_overfit PASSED                                                                                                                       [ 25%]
deepchem/models/tests/test_megnet.py::test_megnet_classification PASSED                                                                                                                [ 50%]
deepchem/models/tests/test_megnet.py::test_megnet_regression PASSED                                                                                                                    [ 75%]
deepchem/models/tests/test_megnet.py::test_megnet_reload PASSED      
```
which is surprising because when I run `pytest -m "not slow" deepchem` I found the same test failing.
```
deepchem/models/tests/test_megnet.py::test_megnet_overfit PASSED                                                                                                                       [ 59%]
deepchem/models/tests/test_megnet.py::test_megnet_classification FAILED                                                                                                                [ 59%]
deepchem/models/tests/test_megnet.py::test_megnet_regression PASSED                                                                                                                    [ 59%]
deepchem/models/tests/test_megnet.py::test_megnet_reload PASSED   
```
Here is the verbose message I obtain with `pytest -m "not slow" deepchem`:
```
========================================================================================== FAILURES ==========================================================================================
_________________________________________________________________________________ test_megnet_classification _________________________________________________________________________________

    @pytest.mark.torch
    def test_megnet_classification():
      fgg = FGG(avg_n_nodes=10,
                n_node_features=5,
                avg_degree=4,
                n_edge_features=3,
                global_features=4,
                num_classes=10)
      graphs = fgg.sample(n_graphs=200)
    
      model = MEGNetModel(n_node_features=5,
                          n_edge_features=3,
                          n_global_features=4,
                          n_blocks=3,
                          is_undirected=True,
                          residual_connection=True,
                          mode='classification',
                          n_classes=10,
                          batch_size=16)
      metric = dc.metrics.Metric(dc.metrics.accuracy_score, mode="classification")
    
>     model.fit(graphs, nb_epoch=50)

deepchem/models/tests/test_megnet.py:67: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
deepchem/models/torch_models/torch_model.py:334: in fit
    return self.fit_generator(
deepchem/models/torch_models/torch_model.py:424: in fit_generator
    outputs = self.model(inputs)
../../anaconda3/envs/deepchem/lib/python3.8/site-packages/torch/nn/modules/module.py:1110: in _call_impl
    return forward_call(*input, **kwargs)
deepchem/models/torch_models/megnet.py:134: in forward
    node_features, edge_features, global_features = self.megnet_blocks[i](
../../anaconda3/envs/deepchem/lib/python3.8/site-packages/torch/nn/modules/module.py:1110: in _call_impl
    return forward_call(*input, **kwargs)
deepchem/models/torch_models/layers.py:1111: in forward
    node_features = self._update_node_features(node_features, edge_index,
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = GraphNetwork(n_node_features=5, n_edge_features=3, n_global_features=4, is_undirected=True, residual_connection=True)
node_features = tensor([[1.8074, 1.1820, 1.6701, 1.4471, 1.9034],
        [1.0560, 1.4975, 1.4365, 1.0258, 1.0910],
        [1.7437, 1..., 0.5697, 0.1883],
        [0.6970, 0.3029, 0.0679, 0.0097, 0.5192],
        [0.6315, 0.9357, 0.5476, 0.9347, 0.2072]])
edge_index = tensor([[  1,   8,   5,  ..., 151, 153, 151],
        [  5,   9,   2,  ..., 150, 152, 157]])
edge_features = tensor([[ 2.1788,  0.9686, -1.7319],
        [ 2.3947,  1.3802, -1.7789],
        [ 2.3693,  1.2649, -1.8082],
       ... 0.2674, -1.1684],
        [ 0.6954,  0.3519, -1.0077],
        [ 0.9760,  0.3110, -1.1428]], grad_fn=<AddmmBackward0>)
global_features = tensor([[1.3261, 1.9068, 1.5379, 1.5821],
        [0.8628, 0.8111, 0.8008, 0.8901],
        [1.9417, 1.8680, 1.1680, 1... [1.5088, 1.9926, 1.1488, 1.6987],
        [0.3604, 0.0077, 0.8755, 0.3739],
        [0.9188, 0.3821, 0.8231, 0.1819]])
batch = tensor([ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  1,  1,  1,  1,  1,  1,  1,  1,
         1,  1,  2,  2,  2,  2,  2,  2...3, 13, 13, 13, 13, 13, 13, 13, 14, 14, 14, 14,
        14, 14, 14, 14, 14, 14, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15])

    def _update_node_features(self, node_features, edge_index, edge_features,
                              global_features, batch):
    
      src_index, dst_index = edge_index
      # Compute mean edge features for each node by dst_index (each node
      # receives information from edges which have that node as its destination,
      # hence the computation uses dst_index to aggregate information)
      edge_features_mean_by_node = scatter_mean(edge_features, dst_index, dim=0)
>     out = torch.cat(
          (node_features, edge_features_mean_by_node, global_features[batch]),
          dim=1)
E     RuntimeError: Sizes of tensors must match except in dimension 1. Expected size 160 but got size 159 for tensor number 1 in the list.

deepchem/models/torch_models/layers.py:1060: RuntimeError
```

**Comment by ojeda-e** [comment]
Any update so far @arunppsg?

**Comment by arunppsg** [comment]
Sorry, I was bit busy. This error is quite peculiar - it raises only when run as a single test but the failure is not captures when run as a collective and also it is falky. I need to take a careful look sometime later.
